File: services/model-catalog-service/src/handlers/adapter.rs
use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
};
use uuid::Uuid;

use crate::AppState;
use crate::models::{
    InferenceContract, ModelAdapter, PublishContractRequest, RegisterAdapterRequest,
};

pub async fn list_adapters(State(state): State<AppState>) -> impl IntoResponse {
    match sqlx::query_as::<_, ModelAdapter>(
        "SELECT * FROM model_adapters ORDER BY created_at DESC",
    )
    .fetch_all(&state.db)
    .await
    {
        Ok(rows) => Json(rows).into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

pub async fn register_adapter(
    State(state): State<AppState>,
    Json(body): Json<RegisterAdapterRequest>,
) -> impl IntoResponse {
    let id = Uuid::now_v7();
    match sqlx::query_as::<_, ModelAdapter>(
        "INSERT INTO model_adapters (id, slug, name, adapter_kind, artifact_uri, sidecar_image, framework, model_id, status) \
         VALUES ($1, $2, $3, $4, $5, $6, $7, $8, 'registered') RETURNING *",
    )
    .bind(id)
    .bind(&body.slug)
    .bind(&body.name)
    .bind(&body.adapter_kind)
    .bind(&body.artifact_uri)
    .bind(&body.sidecar_image)
    .bind(&body.framework)
    .bind(body.model_id)
    .fetch_one(&state.db)
    .await
    {
        Ok(row) => (StatusCode::CREATED, Json(row)).into_response(),
        Err(e) => (StatusCode::BAD_REQUEST, e.to_string()).into_response(),
    }
}

pub async fn get_adapter(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> impl IntoResponse {
    match sqlx::query_as::<_, ModelAdapter>("SELECT * FROM model_adapters WHERE id = $1")
        .bind(id)
        .fetch_optional(&state.db)
        .await
    {
        Ok(Some(row)) => Json(row).into_response(),
        Ok(None) => (StatusCode::NOT_FOUND, "adapter not found").into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

pub async fn list_contracts(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> impl IntoResponse {
    match sqlx::query_as::<_, InferenceContract>(
        "SELECT * FROM inference_contracts WHERE adapter_id = $1 ORDER BY created_at DESC",
    )
    .bind(id)
    .fetch_all(&state.db)
    .await
    {
        Ok(rows) => Json(rows).into_response(),
        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()).into_response(),
    }
}

pub async fn publish_contract(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(body): Json<PublishContractRequest>,
) -> impl IntoResponse {
    let contract_id = Uuid::now_v7();
    match sqlx::query_as::<_, InferenceContract>(
        "INSERT INTO inference_contracts (id, adapter_id, version, input_schema, output_schema) \
         VALUES ($1, $2, $3, $4, $5) RETURNING *",
    )
    .bind(contract_id)
    .bind(id)
    .bind(&body.version)
    .bind(&body.input_schema)
    .bind(&body.output_schema)
    .fetch_one(&state.db)
    .await
    {
        Ok(row) => (StatusCode::CREATED, Json(row)).into_response(),
        Err(e) => (StatusCode::BAD_REQUEST, e.to_string()).into_response(),
    }
}

File: services/model-catalog-service/internal/repo/adapter.go
package repo

import (
	"context"
	"errors"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/openfoundry/openfoundry-go/services/model-catalog-service/internal/models"
)

type AdapterRepo struct {
	Pool *pgxpool.Pool
}

const adapterCols = `id, slug, name, adapter_kind, artifact_uri, sidecar_image,
                     framework, model_id, status, created_at, updated_at`

func scanAdapter(s scanner) (models.ModelAdapter, error) {
	var a models.ModelAdapter
	err := s.Scan(&a.ID, &a.Slug, &a.Name, &a.AdapterKind, &a.ArtifactURI,
		&a.SidecarImage, &a.Framework, &a.ModelID, &a.Status, &a.CreatedAt, &a.UpdatedAt)
	return a, err
}

type scanner interface{ Scan(...any) error }

func (r *AdapterRepo) ListAdapters(ctx context.Context) ([]models.ModelAdapter, error) {
	rows, err := r.Pool.Query(ctx,
		`SELECT `+adapterCols+` FROM model_adapters ORDER BY created_at DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]models.ModelAdapter, 0)
	for rows.Next() {
		a, err := scanAdapter(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, a)
	}
	return out, rows.Err()
}

func (r *AdapterRepo) RegisterAdapter(ctx context.Context, body models.RegisterAdapterRequest) (models.ModelAdapter, error) {
	row := r.Pool.QueryRow(ctx,
		`INSERT INTO model_adapters
                (id, slug, name, adapter_kind, artifact_uri, sidecar_image, framework, model_id, status)
            VALUES ($1, $2, $3, $4, $5, $6, $7, $8, 'registered')
            RETURNING `+adapterCols,
		uuid.New(), body.Slug, body.Name, body.AdapterKind, body.ArtifactURI,
		body.SidecarImage, body.Framework, body.ModelID)
	return scanAdapter(row)
}

func (r *AdapterRepo) GetAdapter(ctx context.Context, id uuid.UUID) (*models.ModelAdapter, error) {
	row := r.Pool.QueryRow(ctx,
		`SELECT `+adapterCols+` FROM model_adapters WHERE id = $1`, id)
	a, err := scanAdapter(row)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &a, nil
}

const contractCols = `id, adapter_id, version, input_schema, output_schema, created_at`

func scanContract(s scanner) (models.InferenceContract, error) {
	var c models.InferenceContract
	err := s.Scan(&c.ID, &c.AdapterID, &c.Version, &c.InputSchema, &c.OutputSchema, &c.CreatedAt)
	return c, err
}

func (r *AdapterRepo) ListContracts(ctx context.Context, adapterID uuid.UUID) ([]models.InferenceContract, error) {
	rows, err := r.Pool.Query(ctx,
		`SELECT `+contractCols+` FROM inference_contracts WHERE adapter_id = $1 ORDER BY created_at DESC`, adapterID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := make([]models.InferenceContract, 0)
	for rows.Next() {
		c, err := scanContract(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, c)
	}
	return out, rows.Err()
}

func (r *AdapterRepo) PublishContract(ctx context.Context, adapterID uuid.UUID, body models.PublishContractRequest) (models.InferenceContract, error) {
	row := r.Pool.QueryRow(ctx,
		`INSERT INTO inference_contracts (id, adapter_id, version, input_schema, output_schema)
           VALUES ($1, $2, $3, $4, $5) RETURNING `+contractCols,
		uuid.New(), adapterID, body.Version, body.InputSchema, body.OutputSchema)
	return scanContract(row)
}
